File: app/src/test/java/com/babestudios/companyinfouk/ui/filinghistorydetails/FilingHistoryDetailsPresenterTest.java
package com.babestudios.companyinfouk.ui.filinghistorydetails;

import android.net.Uri;

import com.babestudios.companyinfouk.data.DataManager;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.junit.MockitoJUnitRunner;

import java.io.File;

import okhttp3.MediaType;
import okhttp3.ResponseBody;
import rx.Observable;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class FilingHistoryDetailsPresenterTest {

	private FilingHistoryDetailsPresenter filingHistoryDetailsPresenter;

	@Before
	public void setUp() {
		filingHistoryDetailsPresenter = new FilingHistoryDetailsPresenter(mock(DataManager.class));
		filingHistoryDetailsPresenter.create();
		FilingHistoryDetailsActivityView view = mock(FilingHistoryDetailsActivityView.class);
		when(view.getFilingHistoryItemString()).thenReturn("0");
		filingHistoryDetailsPresenter.attachView(view);
		when(filingHistoryDetailsPresenter.dataManager.getDocument("0")).thenReturn(Observable.just(ResponseBody.create(MediaType.parse("text/plain"), "test")));
		when(filingHistoryDetailsPresenter.dataManager.writeDocumentPdf(any(ResponseBody.class))).thenReturn(Uri.parse("http:\\some.com"));
	}

	@Test
	public void whenGetDocument_thenDataManagerGetDocumentIsCalled() {
		filingHistoryDetailsPresenter.getDocument();
		verify(filingHistoryDetailsPresenter.dataManager).getDocument("0");
	}

	@Test
	public void whenWritePdf_thenDataManagerWriteDocumentPdfIsCalled() {
		filingHistoryDetailsPresenter.writePdf(ResponseBody.create(MediaType.parse("text/plain"), "test"));
		verify(filingHistoryDetailsPresenter.dataManager).writeDocumentPdf(any(ResponseBody.class));
	}
}

File: app/src/test/java/com/babestudios/companyinfouk/ui/filinghistorydetails/FilingHistoryDetailsPresenterTest.kt
package com.babestudios.companyinfouk.ui.filinghistorydetails

import com.babestudios.companyinfouk.CompaniesHouseApplication
import com.babestudios.companyinfouk.DaggerTestApplicationComponent
import com.babestudios.companyinfouk.TestApplicationModule
import com.babestudios.companyinfouk.data.model.filinghistory.FilingHistoryItem
import com.babestudios.companyinfouk.common.model.filinghistory.FilingHistoryLinksDto
import com.nhaarman.mockitokotlin2.whenever
import io.reactivex.CompletableSource
import io.reactivex.Single
import okhttp3.MediaType
import okhttp3.ResponseBody
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mockito.verify
import org.mockito.junit.MockitoJUnitRunner

@RunWith(MockitoJUnitRunner::class)
class FilingHistoryDetailsPresenterTest {

	private lateinit var filingHistoryDetailsPresenter: FilingHistoryDetailsPresenter

	@Before
	fun setUp() {
		val testApplicationComponent = DaggerTestApplicationComponent.builder()
				.testApplicationModule(TestApplicationModule(CompaniesHouseApplication()))
				.build()
		filingHistoryDetailsPresenter = testApplicationComponent.filingHistoryDetailsPresenter()
		val historyLinks = FilingHistoryLinksDto()
		historyLinks.documentMetadata = "something"
		val filingHistoryItem = FilingHistoryItem()
		filingHistoryItem.links = historyLinks
		val filingHistoryDetailsViewModel = FilingHistoryDetailsViewModel()
		filingHistoryDetailsViewModel.state.value.filingHistoryItem = filingHistoryItem
		filingHistoryDetailsPresenter.setViewModel(filingHistoryDetailsViewModel, CompletableSource {  })
		whenever(filingHistoryDetailsPresenter.companiesRepository.getDocument("something")).thenReturn(Single.just(ResponseBody.create(MediaType.parse("text/plain"), "test")))
	}

	@Test
	fun whenGetDocument_thenDataManagerGetDocumentIsCalled() {
		filingHistoryDetailsPresenter.fetchDocument()
		verify(filingHistoryDetailsPresenter.companiesRepository)?.getDocument("something")
	}

	/**
	 * TODO This goes through the Activity due to permission requests. Rewrite as an Espresso test?
	 */
	/*@Test
	fun whenWritePdf_thenDataManagerWriteDocumentPdfIsCalled() {
		filingHistoryDetailsPresenter.writeDocument()
		verify(filingHistoryDetailsPresenter.companiesRepository).writeDocumentPdf(any())
	}*/
}
